File: src/go/ch01/1.3/l2/totient.go
package main

import "errors"

// PhiU64 computes Euler’s totient φ(n) for n>=1 using trial factorization.
func PhiU64(n uint64) (uint64, error) {
	if n == 0 {
		return 0, errors.New("n must be positive")
	}
	result := n
	x := n
	var p uint64 = 2
	for p*p <= x {
		if x%p == 0 {
			for x%p == 0 {
				x /= p
			}
			result -= result / p
		}
		if p == 2 {
			p = 3
		} else {
			p += 2
		}
	}
	if x > 1 {
		result -= result / x
	}
	return result, nil
}

File: src/erlang/ch01/1.3/l2/src/totient.erl
-module(totient).
-export([phi/1]).

%% φ(N): count of 1..N coprime to N. Teaching version via trial factorization.
phi(N) when N =< 0 ->
    erlang:error(badarg);
phi(N) ->
    phi_(N, N, 2).

phi_(X, Result, P) when P*P > X ->
    case X > 1 of
        true  -> Result - (Result div X);
        false -> Result
    end;
phi_(X, Result, 2) ->
    case X rem 2 of
        0 -> phi_(strip(X, 2), Result - (Result div 2), 3);
        _ -> phi_(X, Result, 3)
    end;
phi_(X, Result, P) ->
    case X rem P of
        0 -> phi_(strip(X, P), Result - (Result div P), P+2);
        _ -> phi_(X, Result, P+2)
    end.

strip(X, P) when X rem P =:= 0 -> strip(X div P, P);
strip(X, _P) -> X.